pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java

```diff
@@ -86,19 +90,29 @@ public int submit(long requestId, QueryPlan queryPlan, long timeoutMs)
           String host = serverInstance.getHostname();
           int servicePort = serverInstance.getQueryServicePort();
           DispatchClient client = getOrCreateDispatchClient(host, servicePort);
-          Worker.QueryResponse response = client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
+          Future<Worker.QueryResponse> response = client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
                   QueryPlanSerDeUtils.serialize(constructDistributedStagePlan(queryPlan, stageId, serverInstance)))
               .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_ID, String.valueOf(requestId))
               .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs)).build());
-          if (response.containsMetadata(QueryConfig.KEY_OF_SERVER_RESPONSE_STATUS_ERROR)) {
-            throw new RuntimeException(
-                String.format("Unable to execute query plan at stage %s on server %s: ERROR: %s", stageId,
-                    serverInstance, response));
-          }
+          querySubmitResponse.add(response);
+        }
+      }
+    }
+    while (System.currentTimeMillis() < submissionDeadlineMs) {
+      for (Future<Worker.QueryResponse> future : querySubmitResponse) {
+        if (!future.isDone()) {
+          break;
+          Thread.sleep(100);
```

Comment:
Is this misplaced ? You want to break and then sleep before checking again right ?

Reply:
ideally what's the best handling for this is similar to what combineOperator submits tasks. e.g. use a countdownlatch to ensure all futures returns properly and check on the countdownlatch.

Reply:
First of all, adding RPC has nothing to do with blocking stub and aysnc stub.

Second, we shouldn't use busy waiting loop waiting for future result. Can you either chain the future or use latch to wait for the result asynchronously?